src/main/java/sparksoniq/spark/iterator/flowr/OrderByClauseSparkIterator.java

```diff
@@ -17,46 +17,149 @@
+     * Requires _child iterator to be opened.
+     * @return Sorted TreeMap(ascending). key - atomics from expressions, value - input tuples
+     */
+    private TreeMap<FlworKey,  List<FlworTuple>> mapExpressionsToOrderedPairs() {
+        // tree map keeps the natural item order deduced from an implementation of Comparator
+        // OrderByClauseSortClosure implements a comparator and provides the exact desired behavior for local execution as well
+        TreeMap<FlworKey, List<FlworTuple>> keyValuePairs = new TreeMap<>(new OrderByClauseSortClosure(_expressions, true));
+
+        // assign current context as parent. re-use the same context object for efficiency
+        DynamicContext tupleContext = new DynamicContext(_currentDynamicContext);
+        while (_child.hasNext()) {
+            FlworTuple inputTuple = _child.next();
+
+            List<Item> results = new ArrayList<>(); // results from the expressions will become a key
+            for (OrderByClauseSparkIteratorExpression orderByExpression : _expressions) {
+                tupleContext.removeAllVariables();                     // clear the previous variables
+                tupleContext.setBindingsFromTuple(inputTuple);        // assign new variables from new tuple
+
+                RuntimeIterator expression = orderByExpression.getExpression();
+                expression.open(tupleContext);
+                while (expression.hasNext()) {
+                    Item resultItem = expression.next();
+                    if (resultItem != null && !Item.isAtomic(resultItem))
+                        throw new NonAtomicKeyException("Order by keys must be atomics",
+                                orderByExpression.getIteratorMetadata().getExpressionMetadata());
+                    results.add(resultItem);
+                }
+                expression.close();
+            }
+            FlworKey key = new FlworKey(results);
+            List<FlworTuple> values = keyValuePairs.get(key);   // all values for a single matching key are held in a list
+            if (values == null) {
+                values = new ArrayList<>();
+                keyValuePairs.put(key, values);
+            }
+            values.add(inputTuple);
+        }
+        return keyValuePairs;
     }
 
     @Override
     public JavaRDD<FlworTuple> getRDD(DynamicContext context) {
-        if(this._rdd == null) {
-            this._rdd = this._child.getRDD(context);
-            //map to pairs - ArrayItem [sort keys] , tuples
-            JavaPairRDD<FlworKey, FlworTuple> keyTuplePair = this._rdd
-                    .mapToPair(new OrderByMapToPairClosure(this._expressions, _isStable));
-            //sort by key
-            keyTuplePair = keyTuplePair.sortByKey(new OrderByClauseSortClosure(this._expressions, _isStable));
-            //map back to tuple RDD
-            this._rdd = keyTuplePair.map(tuple2 -> tuple2._2());
+        if (this._rdd == null) {
```

Comment:
I think we are no longer caching RDDs. Since they depend on the context, this is now incorrect.

Reply:
Yes, I missed this, it will be fixed.